File: src/main/java/org/parthvnp/DynamicProgramming/HouseRobber.java
package org.parthvnp.DynamicProgramming;

public class HouseRobber {
	public int rob(int[] nums) {
		int[] dp = new int[nums.length];
		if (nums.length == 1) return nums[0];
		dp[0] = nums[0];
		dp[1] = Math.max(nums[0], nums[1]);
		for(int i = 2; i < nums.length; i++) dp[i] = Math.max(dp[i-1], dp[i-2] + nums[i]);
		return dp[nums.length - 1];
	}

  public static void main(String[] args) {
	  var s = new HouseRobber();
	  var e1 = s.rob(new int[] {1, 2, 3, 1});
	  System.out.println(e1 == 4);
  }
}

File: src/main/python/dynamic_programming/house_robber.py
from typing import *
from shared.test import test


class HouseRobber:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return nums[0]
        dp = [0] * n
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2,n): dp[i] = max(dp[i-1], dp[i-2] + nums[i])
        return dp[n - 1]

    def rob_top_down(self, nums: List[int]) -> int:
        n  = len(nums)
        memo = {}
        def dp(i):
            if i == 0 : return nums[0]
            if i == 1: return max(nums[0], nums[1])
            if i not in memo: memo[i] = max(dp(i-1), dp(i-2) + nums[i])
            return memo[i]
        return dp(n - 1)

if __name__ == '__main__':
    s = HouseRobber()
    e1 = s.rob([1,2,3,1])
    test(e1, 4)
    e2 = s.rob_top_down([1,2,3,1])
    test(e2, 4)
